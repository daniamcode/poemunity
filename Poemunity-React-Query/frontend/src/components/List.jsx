import React, { useEffect, useState, useContext } from 'react'
import { Link } from 'react-router-dom'
import usePoems from '../react-query/usePoems'
import useDeletePoem from '../react-query/useDeletePoem'
import useLikePoem from '../react-query/useLikePoem'
import { AppContext } from '../App';
import './List.scss'
import './Detail.scss'
import '../App.scss'
import { TextField } from '@material-ui/core'
import SearchIcon from '@material-ui/icons/Search'
import HighlightOffSharpIcon from '@material-ui/icons/HighlightOffSharp'
import EditIcon from '@material-ui/icons/Edit';
import SubjectSharpIcon from '@material-ui/icons/SubjectSharp'
import CircularProgress from './CircularIndeterminate'
import capitalizeFirstLetter from '../utils/capitalizeFirstLetter'
import sortPoems from '../utils/sortPoems'
import { Helmet } from 'react-helmet'
import {
  LIKE,
  LIKES,
  READ_MORE,
  SEARCH_PLACEHOLDER,
  ORDER_BY,
  ORDER_BY_DATE,
  ORDER_BY_LIKES,
  ORDER_BY_RANDOM,
  ORDER_BY_TITLE,
  CATEGORIES_TITLE_LABEL
} from '../data/constants'
import normalizeString from '../utils/normalizeString'
import { useHistory } from "react-router-dom";
import { addQueryParam, useFiltersFromQuery } from '../utils/urlUtils.js'

function List (props) {
  const genre = props.match.params.genre
  const [poems, setPoems] = useState([])
  const [filter, setFilter] = useState('')
  
  const history = useHistory();
  const context = useContext(AppContext);
  const poemsQuery = usePoems()

  console.log(poemsQuery)

  const [paramsData, setParamsData] = useFiltersFromQuery({
    orderBy: null
  })

  useEffect(()=> {
    if(poemsQuery.data) {
      const newData = [...poemsQuery.data]
      
      if(genre) {
        const poemsFiltered = newData.filter((poems) => poems.genre === genre)
        const poemsSorted = sortPoems(paramsData.orderBy, poemsFiltered)
        setPoems(poemsSorted)
      }
      else {
        const poemsSorted = sortPoems(paramsData.orderBy, newData)
        setPoems(poemsSorted)
      }
    }
  }, [JSON.stringify([poemsQuery.data, genre, paramsData])])
  

  const handleOrderChange = (event) => {
    addQueryParam({id: 'orderBy', value: event.target.value})
    setParamsData({orderBy: event.target.value})
  }

  const deletePoemMutation = useDeletePoem()
  const likePoemMutation = useLikePoem()

  const onLike = (event, poemId) => {
    event.preventDefault()
    likePoemMutation.mutate(poemId)
  }


  const editPoem = (poemId) => {
    const newPath = '/profile'
    history.push(newPath);
    context.setState({elementToEdit: poemId})
  }

  const handleSearchChange = (event) => {
    setFilter(normalizeString(event.target.value))
  }

  if (poemsQuery.isLoading) {
    return <CircularProgress />
  }

  return (
    <>
      <Helmet>
        <title>
          {genre ? `${capitalizeFirstLetter(genre)} poems` : 'Poemunity'}
        </title>
      </Helmet>
      <div className='list__container'>
        <div className='list__intro'>
          {genre && (
            <p className='list__presentation'>
              {CATEGORIES_TITLE_LABEL}
              {genre.toUpperCase()}
            </p>
          )}
          <div className='list__search'>
          <div className='separator' />
            <SearchIcon style={{ fontSize: 40, fill: '#4F5D73' }} />
            <TextField
              label={SEARCH_PLACEHOLDER}
              InputLabelProps={{
                style: { color: '#4F5D73' }
              }}
              InputProps={{
                style: { color: '#4F5D73' }
              }}
              onChange={handleSearchChange}
            />
          </div>
          <form className='list__sort'>
            <label>
              {ORDER_BY}
              <select
                type='submit'
                id='sort'
                name='sort'
                onChange={handleOrderChange}
              >
                <option value={ORDER_BY_LIKES} selected={ORDER_BY_LIKES === paramsData.orderBy}>{ORDER_BY_LIKES}</option>
                <option value={ORDER_BY_DATE} selected={ORDER_BY_DATE === paramsData.orderBy}>{ORDER_BY_DATE}</option>
                <option value={ORDER_BY_RANDOM} selected={ORDER_BY_RANDOM === paramsData.orderBy}>{ORDER_BY_RANDOM}</option>
                <option value={ORDER_BY_TITLE} selected={ORDER_BY_TITLE === paramsData.orderBy}>{ORDER_BY_TITLE}</option>
              </select>
            </label>
          </form>
        </div>

        {poems?.map((poem) => (
          <main key={poem.id} className='poem__detail'>
            {normalizeString(poem.author).includes(filter) && (
              <section className='poem__block'>
                <section>
                  <Link to={`/detail/${poem.id}`} className='poem__title'>
                    {poem.title}
                  </Link>
                  <div className='poem__author-container'>
                    <img className='poem__picture' src={poem.picture} />
                    <p className='poem__author'>{poem.author}</p>
                  </div>
                  <div className='poem__date'>{poem.date}</div>
                </section>
                <section>
                  <div className='poem__content poems__content'>
                    {poem.poem}
                  </div>
                  <div className='poems__read-more'>
                    <Link
                      to={`/detail/${poem.id}`}
                      className='poems__read-more'
                    >
                      {READ_MORE}
                    </Link>
                  </div>
                </section>
                <section className='poem__footer'>
                  {poem.likes.length === 1 && (
                    <div className='poem__likes'>
                      {poem.likes.length} {LIKE}
                    </div>
                  )}
                  {poem.likes.length !== 1 && (
                    <div className='poem__likes'>
                      {poem.likes.length} {LIKES}
                    </div>
                  )}
                  <div className='separator' />
                  {context.user &&
                    poem.author !== context.username &&
                    poem.likes.some((id) => id === context.username) && (
                      <Link
                        className='poem__likes-icon'
                        onClick={(event) => onLike(event, poem.id)}
                      />
                  )}
                  {context.user &&
                    poem.author !== context.username &&
                    !poem.likes.some((id) => id === context.username) && (
                      <Link
                        className='poem__unlikes-icon'
                        onClick={(event) => onLike(event, poem.id)}
                      />
                  )}
                  {context.user &&
                    (poem.author === context.username ||
                      context.userId === context.adminId) && (
                    <EditIcon
                    className='poem__edit-icon'
                    onClick={(event) => editPoem(poem.id)}
                  />
                  )}
                  {context.user &&
                    (poem.author === context.username ||
                      context.userId === context.adminId) && (
                        <HighlightOffSharpIcon
                          className='poem__delete-icon'
                          style={{ fill: 'red' }}
                          onClick={(event) => deletePoemMutation.mutate(poem.id)}
                        />
                  )}
                  <Link
                    to={`/detail/${poem.id}`}
                    className='poem__comments-icon'
                  >
                    <SubjectSharpIcon style={{ fill: '#000' }} />
                  </Link>
                </section>
              </section>
            )}
          </main>
        ))}
      </div>
    </>
  )
}
export default List
