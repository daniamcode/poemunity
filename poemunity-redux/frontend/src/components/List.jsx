import React, { useEffect, useState, useContext } from 'react'
import { Link, useHistory } from 'react-router-dom'
import usePoems from '../react-query/usePoems'
import useDeletePoem from '../react-query/useDeletePoem'
import useLikePoem from '../react-query/useLikePoem'
import { AppContext } from '../App'
import './List.scss'
import './Detail.scss'
import '../App.scss'
import { TextField } from '@material-ui/core'
import SearchIcon from '@material-ui/icons/Search'
import HighlightOffSharpIcon from '@material-ui/icons/HighlightOffSharp'
import EditIcon from '@material-ui/icons/Edit'
import SubjectSharpIcon from '@material-ui/icons/SubjectSharp'
import CircularProgress from './CircularIndeterminate'
import capitalizeFirstLetter from '../utils/capitalizeFirstLetter'
import sortPoems from '../utils/sortPoems'
import { Helmet } from 'react-helmet'
import {
  LIKE,
  LIKES,
  READ_MORE,
  SEARCH_PLACEHOLDER,
  ORDER_BY,
  ORDER_BY_DATE,
  ORDER_BY_LIKES,
  ORDER_BY_RANDOM,
  ORDER_BY_TITLE,
  CATEGORIES_TITLE_LABEL
} from '../data/constants'
import normalizeString from '../utils/normalizeString'
import { addQueryParam, useFiltersFromQuery } from '../utils/urlUtils.js'
import { strings, arrays, dom, objects } from '@daniamcode/utils'
import ListItem from './ListItem'
import { useDispatch, useSelector } from 'react-redux'
import { getPoemsAction } from '../redux/actions/poemActions'

function List (props) {
  const genre = props.match.params.genre
  const [poems, setPoems] = useState([])
  const [filter, setFilter] = useState('')

  const [paramsData, setParamsData] = useFiltersFromQuery({
    orderBy: null,
    origin: 'all'
  })
  
  const history = useHistory()
  const context = useContext(AppContext)

  // Redux
  const dispatch = useDispatch();

  const {
      poemsQuery,
  } = useSelector(state => state);

  useEffect(() => {
      dispatch(getPoemsAction({}))
  }, []);

  useEffect(() => {
    if (poemsQuery && poemsQuery.item && poemsQuery.item.length > 0) {
      const newData = [...poemsQuery.item]

      if (genre) {
        const poemsFiltered = newData.filter((poems) => poems.genre === genre)
        const poemsSorted = sortPoems(paramsData.orderBy, poemsFiltered)
        setPoems(poemsSorted)
      } else {
        const poemsSorted = sortPoems(paramsData.orderBy, newData)
        setPoems(poemsSorted)
      }
    }
  }, [JSON.stringify([poemsQuery, genre, paramsData])])

  const handleOrderChange = (event) => {
    addQueryParam({ id: 'orderBy', value: event.target.value })
    setParamsData({ ...paramsData, orderBy: event.target.value })
  }

  const handleOriginChange = (event) => {
    addQueryParam({ id: 'origin', value: event.target.value })
    setParamsData({ ...paramsData, origin: event.target.value })
  }

  const deletePoemMutation = useDeletePoem()
  const likePoemMutation = useLikePoem()

  const onLike = (event, poemId) => {
    event.preventDefault()
    likePoemMutation.mutate(poemId)
  }

  const editPoem = (poemId) => {
    const newPath = '/profile'
    history.push(newPath)
    context.setState({ ...context, elementToEdit: poemId })
  }

  const handleSearchChange = (event) => {
    setFilter(normalizeString(event.target.value))
  }

  if (poemsQuery.isFetching) {
    console.log('test my first npm package!')
    console.log(strings.isLowercase('Qwerty'))
    console.log(strings.isLowercase('qwerty'))
    console.log(strings.isLowercase('ABC'))
    return <CircularProgress />
  }

  return (
    <>
      <Helmet>
        <title>
          {genre ? `${capitalizeFirstLetter(genre)} poems` : 'Poemunity'}
        </title>
      </Helmet>
      <div className='list__container'>
        <div className='list__intro'>
          {genre && (
            <p className='list__presentation'>
              {CATEGORIES_TITLE_LABEL}
              {genre.toUpperCase()}
            </p>
          )}
          <div className='list__search'>
            <div className='separator' />
            <SearchIcon style={{ fontSize: 40, fill: '#4F5D73' }} />
            <TextField
              label={SEARCH_PLACEHOLDER}
              InputLabelProps={{
                style: { color: '#4F5D73' }
              }}
              InputProps={{
                style: { color: '#4F5D73' }
              }}
              onChange={handleSearchChange}
            />
          </div>
          <form className='list__sort'>
            <label>
              Authors:
              <select
                type='submit'
                id='origin'
                name='origin'
                onChange={handleOriginChange}
              >
                <option value='all' selected={'all' === paramsData.origin}>All</option>
                <option value='famous' selected={'famous' === paramsData.origin}>Famous</option>
                <option value='user' selected={'user' === paramsData.origin}>Users</option>
              </select>
            </label>
          </form>
          <form className='list__sort'>
            <label>
              {ORDER_BY}
              <select
                type='submit'
                id='sort'
                name='sort'
                onChange={handleOrderChange}
              >
                <option value={ORDER_BY_LIKES} selected={ORDER_BY_LIKES === paramsData.orderBy}>{ORDER_BY_LIKES}</option>
                <option value={ORDER_BY_DATE} selected={ORDER_BY_DATE === paramsData.orderBy}>{ORDER_BY_DATE}</option>
                <option value={ORDER_BY_RANDOM} selected={ORDER_BY_RANDOM === paramsData.orderBy}>{ORDER_BY_RANDOM}</option>
                <option value={ORDER_BY_TITLE} selected={ORDER_BY_TITLE === paramsData.orderBy}>{ORDER_BY_TITLE}</option>
              </select>
            </label>
          </form>
        </div>

        {poems.map((poem) => (
          <ListItem 
            poem={poem}
            filter={filter}
            context={context}
          />
        ))}
      </div>
    </>
  )
}
export default List
