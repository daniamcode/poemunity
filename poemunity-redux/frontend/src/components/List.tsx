import { useEffect, useState, useContext } from 'react'
import { AppContext } from '../App'
import './List.scss'
import './Detail.scss'
import '../App.scss'
import { TextField } from '@material-ui/core'
import SearchIcon from '@material-ui/icons/Search'
import CircularProgress from './CircularIndeterminate'
import capitalizeFirstLetter from '../utils/capitalizeFirstLetter'
import sortPoems from '../utils/sortPoems'
import { Helmet } from 'react-helmet'
import {
  SEARCH_PLACEHOLDER,
  ORDER_BY,
  ORDER_BY_DATE,
  ORDER_BY_LIKES,
  ORDER_BY_RANDOM,
  ORDER_BY_TITLE,
  CATEGORIES_TITLE_LABEL
} from '../data/constants'
import normalizeString from '../utils/normalizeString'
import { addQueryParam, useFiltersFromQuery } from '../utils/urlUtils'
// import { strings, arrays, dom, objects } from '@daniamcode/utils'
import ListItem from './ListItem'
import { useSelector } from 'react-redux'
import { useAppDispatch, RootState } from '../redux/store'
import { getPoemsListAction } from '../redux/actions/poemsActions'
import { Poem } from '../typescript/interfaces'



interface ListProps {
  match: {
    params: {
      genre: string
    }
  }
}

function List (props: ListProps) {
  interface ListStates {
    poems: Poem[]
    filter: string
  }
  const genre = props.match.params.genre
  const [poems, setPoems] = useState<ListStates["poems"]>([])
  const [filter, setFilter] = useState<ListStates["filter"]>('')

  const [paramsData, setParamsData] = useFiltersFromQuery({
    orderBy: '',
    origin: 'all'
  })
  
  const context = useContext(AppContext)

  // Redux
  const dispatch = useAppDispatch();

  const {
      poemsListQuery
  } = useSelector((state: RootState) => state)

  useEffect(() => {
    const queryOptions = {
        reset: true,
        fetch: false,
    };
    dispatch(getPoemsListAction({
        options: queryOptions,
        params: {}
    }));
  }, []);

  function handleLoadPoems() {
      if (paramsData.origin) {
          const queryOptions = {
              reset: true,
              fetch: true,
          };
          dispatch(getPoemsListAction({
              params: paramsData.origin !== 'all' ? {origin: paramsData.origin} : null,
              options: queryOptions,
          }));
      }
  }

  useEffect(() => {
      handleLoadPoems();
  }, [JSON.stringify(paramsData.origin)]);

  useEffect(() => {
    if (poemsListQuery && poemsListQuery.item && poemsListQuery.item.length > 0) {
      const newData = [...poemsListQuery.item]

      if (genre) {
        const poemsFiltered = newData.filter((poems) => poems.genre === genre)
        const poemsSorted = sortPoems(paramsData.orderBy, poemsFiltered)
        setPoems(poemsSorted)
      } else {
        const poemsSorted = sortPoems(paramsData.orderBy, newData)
        setPoems(poemsSorted)
      }
    }
  }, [JSON.stringify([poemsListQuery, genre, paramsData])])

  const handleOrderChange = (event:React.ChangeEvent<HTMLSelectElement>) => {
    addQueryParam({ id: 'orderBy', value: event.target.value })
    setParamsData({ ...paramsData, orderBy: event.target.value })
  }

  const handleOriginChange = (event:React.ChangeEvent<HTMLSelectElement>) => {
    addQueryParam({ id: 'origin', value: event.target.value })
    setParamsData({ ...paramsData, origin: event.target.value })
  }

  const handleSearchChange = (event:React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setFilter(normalizeString(event.target.value))
  }

  if (poemsListQuery?.isFetching) {
    // console.log('test my first npm package!')
    // console.log(strings.isLowercase('Qwerty'))
    // console.log(strings.isLowercase('qwerty'))
    // console.log(strings.isLowercase('ABC'))
    return <CircularProgress />
  }

  return (
    <>
      <Helmet>
        <title>
          {genre ? `${capitalizeFirstLetter(genre)} poems` : 'Poemunity'}
        </title>
      </Helmet>
      <div className='list__container'>
        <div className='list__intro'>
          {genre && (
            <p className='list__presentation'>
              {CATEGORIES_TITLE_LABEL}
              {genre.toUpperCase()}
            </p>
          )}
          <div className='list__search'>
            <div className='separator' />
            <SearchIcon style={{ fontSize: 40, fill: '#4F5D73' }} />
            <TextField
              label={SEARCH_PLACEHOLDER}
              InputLabelProps={{
                style: { color: '#4F5D73' }
              }}
              InputProps={{
                style: { color: '#4F5D73' }
              }}
              onChange={handleSearchChange}
            />
          </div>
          <form className='list__sort'>
            <label>
              Authors:
              <select
                id='origin'
                name='origin'
                onChange={handleOriginChange}
              >
                <option value='all' selected={'all' === paramsData.origin}>All</option>
                <option value='famous' selected={'famous' === paramsData.origin}>Famous</option>
                <option value='user' selected={'user' === paramsData.origin}>Users</option>
              </select>
            </label>
          </form>
          <form className='list__sort'>
            <label>
              {ORDER_BY}
              <select
                id='sort'
                name='sort'
                onChange={handleOrderChange}
              >
                <option value={ORDER_BY_LIKES} selected={ORDER_BY_LIKES === paramsData.orderBy}>{ORDER_BY_LIKES}</option>
                <option value={ORDER_BY_DATE} selected={ORDER_BY_DATE === paramsData.orderBy}>{ORDER_BY_DATE}</option>
                <option value={ORDER_BY_RANDOM} selected={ORDER_BY_RANDOM === paramsData.orderBy}>{ORDER_BY_RANDOM}</option>
                <option value={ORDER_BY_TITLE} selected={ORDER_BY_TITLE === paramsData.orderBy}>{ORDER_BY_TITLE}</option>
              </select>
            </label>
          </form>
        </div>

        {poems.map((poem) => (
          <ListItem 
            poem={poem}
            filter={filter}
            context={context}
          />
        ))}
      </div>
    </>
  )
}
export default List
