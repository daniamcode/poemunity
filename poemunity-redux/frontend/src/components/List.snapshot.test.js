import React from 'react'
import renderer from 'react-test-renderer'
import List from './List'
import { BrowserRouter } from 'react-router-dom'
import {
  QueryClient,
  QueryClientProvider,
} from "react-query"
import configureStore from '../redux/store';
import { Provider } from 'react-redux';

function renderList (arg) {
  const defaultProps = {
    match: {
      params: {}
    }
  }

  const props = { ...defaultProps, ...arg }
  const queryClient = new QueryClient();

  const store = configureStore();

  return renderer.create(
    <Provider store={store}>
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <List {...props} />
      </BrowserRouter>
    </QueryClientProvider>
    </Provider>
  )
}

describe('List', () => {
  let ListTree

  beforeEach(async () => {
    ListTree = renderList()
  })

  it('should match without id', async () => {
    expect(ListTree).toMatchSnapshot()
  })
})
