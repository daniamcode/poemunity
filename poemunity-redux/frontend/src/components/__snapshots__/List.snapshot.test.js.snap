// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should match without id 1`] = `
<div
  className="list__container"
>
  <div
    className="list__intro"
  >
    <div
      className="list__search"
    >
      <div
        className="separator"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        style={
          Object {
            "fill": "#4F5D73",
            "fontSize": 40,
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <div
        className="MuiFormControl-root MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          style={
            Object {
              "color": "#4F5D73",
            }
          }
        >
          Search an author
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
          style={
            Object {
              "color": "#4F5D73",
            }
          }
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
        </div>
      </div>
    </div>
    <form
      className="list__sort"
    >
      <label>
        Authors:
        <select
          id="origin"
          name="origin"
          onChange={[Function]}
        >
          <option
            selected={true}
            value="all"
          >
            All
          </option>
          <option
            selected={false}
            value="famous"
          >
            Famous
          </option>
          <option
            selected={false}
            value="user"
          >
            Users
          </option>
        </select>
      </label>
    </form>
    <form
      className="list__sort"
    >
      <label>
        Order poems by: 
        <select
          id="sort"
          name="sort"
          onChange={[Function]}
        >
          <option
            selected={false}
            value="Likes"
          >
            Likes
          </option>
          <option
            selected={false}
            value="Date"
          >
            Date
          </option>
          <option
            selected={false}
            value="Random"
          >
            Random
          </option>
          <option
            selected={false}
            value="Title"
          >
            Title
          </option>
        </select>
      </label>
    </form>
  </div>
</div>
`;
